src/Rules/Classes/ClassAttributesRule.php

```diff
@@ -28,12 +31,23 @@ public function processNode(Node $node, Scope $scope): array
 	{
 		$classLikeNode = $node->getOriginalNode();
 
-		return $this->attributesCheck->check(
+		$errors = $this->attributesCheck->check(
 			$scope,
 			$classLikeNode->attrGroups,
 			Attribute::TARGET_CLASS,
 			'class',
 		);
+
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection !== null && $classReflection->isReadOnly()) {
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Cannot apply #[AllowDynamicProperties] to readonly class %s.', $classReflection->getName()))
```

Comment:
We don't use this error message format anywhere. I'd follow some error from AttributesCheck. Example:

> Attribute class AllowDynamicProperties cannot be used with readonly classes.